halo2_proofs/src/circuit/floor_planner/v1.rs

```diff
@@ -321,7 +321,7 @@ impl<'p, 'a, F: Field, CS: Assignment<F> + 'a> AssignmentPass<'p, 'a, F, CS> {
                     _ => None,
                 }) {
                 Some(Some(len)) => len,
-                _ => return Err(Error::Synthesis), // TODO better error
+                _ => return Err(Error::TableCellsNotAssigned),
```

Comment:
This code is duplicated between layouter implementations that have a `default_and_assigned` field. If it were implemented as a helper function that takes `default_and_assigned` as input and returns a `Result<usize, Error>` for the length, then it could be deduplicated, and also that would make it worthwhile to spend more effort collecting error information.

I.e. something like
```rust
let first_unused = assigned_length(default_and_assigned)?;
```
where
```rust
fn assigned_length(
    default_and_assigned: &HashMap<TableColumn, (DefaultTableValue<F>, Vec<bool>)>
) -> Result<usize, Error> {
    let compute_lengths = || default_and_assigned
        .values()
        .map(|(_, assigned)| {
            if assigned.iter().all(|b| *b) {
                Some(assigned.len())
            } else {
                None
            }
        });
    match compute_lengths()
        .reduce(|acc, item| match (acc, item) {
            (Some(a), Some(b)) if a == b => Some(a),
            _ => None,
        }) {
        Some(Some(len)) => len,
        _ => {
            let error_info = ...; // based on the result of compute_lengths()
            Err(Error::TableCellsNotAssigned::new(error_info))
        }
    }
}
```

Reply:
I've tried out this approach in https://github.com/zcash/halo2/pull/755.